File: app/src/main/java/com/mytlogos/enterprise/viewmodel/ListsViewModel.java
package com.mytlogos.enterprise.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.mytlogos.enterprise.background.TaskManager;
import com.mytlogos.enterprise.model.MediaList;
import com.mytlogos.enterprise.model.MediaListSetting;
import com.mytlogos.enterprise.model.MediumSetting;
import com.mytlogos.enterprise.model.ToDownload;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class ListsViewModel extends RepoViewModel {

    private LiveData<? extends MediaListSetting> listSettings;
    private LiveData<MediumSetting> settings;

    public ListsViewModel(@NonNull Application application) {
        super(application);
    }

    public LiveData<List<MediaList>> getLists() {
        return repository.getLists();
    }

    public LiveData<List<MediaList>> getInternLists() {
        return repository.getInternLists();
    }

    public LiveData<? extends MediaListSetting> getListSettings(int id, boolean isExternal) {
        if (this.listSettings == null) {
            listSettings = repository.getListSettings(id, isExternal);
        }
        return listSettings;
    }

    public CompletableFuture<String> updateListName(MediaListSetting listSetting, String text) {
        return repository.updateListName(listSetting, text);
    }

    public CompletableFuture<String> updateListMedium(MediaListSetting listSetting, int newMediumType) {
        return repository.updateListMedium(listSetting, newMediumType);
    }

    public void updateToDownload(boolean add, ToDownload toDownload) {
        TaskManager.runTask(() -> repository.updateToDownload(add, toDownload));
    }

    public LiveData<MediumSetting> getMediumSettings(int mediumId) {
        if (this.settings == null) {
            this.settings = repository.getMediumSettings(mediumId);
        }
        return this.settings;
    }

    public CompletableFuture<String> updateMedium(MediumSetting mediumSettings) {
        return repository.updateMediumType(mediumSettings);
    }

    public CompletableFuture<Boolean> moveMediumToList(int oldListId, int newListId, Collection<Integer> ids) {
        return repository.moveMediaToList(oldListId, newListId, ids);
    }

    public CompletableFuture<Boolean> addMediumToList(int listId, Collection<Integer> ids) {
        return repository.addMediumToList(listId, ids);
    }
}

File: app/src/main/java/com/mytlogos/enterprise/viewmodel/ListsViewModel.kt
package com.mytlogos.enterprise.viewmodel

import android.app.Application
import androidx.lifecycle.LiveData
import com.mytlogos.enterprise.background.TaskManager.Companion.runTask
import com.mytlogos.enterprise.background.repository.MediaListRepository
import com.mytlogos.enterprise.background.repository.MediumRepository
import com.mytlogos.enterprise.model.MediaList
import com.mytlogos.enterprise.model.MediaListSetting
import com.mytlogos.enterprise.model.MediumSetting
import com.mytlogos.enterprise.model.ToDownload
import java.util.concurrent.CompletableFuture

class ListsViewModel(application: Application) : RepoViewModel(application) {
    private var listSettings: LiveData<out MediaListSetting?>? = null
    private var settings: LiveData<MediumSetting>? = null

    val lists: LiveData<MutableList<MediaList>>
        get() = repository.lists

    val internLists: LiveData<MutableList<MediaList>>
        get() = repository.internLists

    private val mediaListRepository by lazy { MediaListRepository.getInstance(application) }
    private val mediumRepository by lazy { MediumRepository.getInstance(application) }

    fun getListSettings(id: Int, isExternal: Boolean): LiveData<out MediaListSetting?> {
        if (listSettings == null) {
            listSettings = repository.getListSettings(id, isExternal)
        }
        return listSettings!!
    }

    suspend fun updateListName(listSetting: MediaListSetting, text: String): String {
        return repository.updateListName(listSetting, text)
    }

    suspend fun updateListMedium(
        listSetting: MediaListSetting?,
        newMediumType: Int
    ): String {
        return repository.updateListMedium(listSetting!!, newMediumType)
    }

    fun updateToDownload(add: Boolean, toDownload: ToDownload?) {
        runTask { repository.updateToDownload(add, toDownload!!) }
    }

    fun getMediumSettings(mediumId: Int): LiveData<MediumSetting> {
        if (settings == null) {
            settings = repository.getMediumSettings(mediumId)
        }
        return settings!!
    }

    suspend fun updateMedium(mediumSettings: MediumSetting): String {
        return mediumRepository.updateMedium(mediumSettings)
    }

    suspend fun moveMediumToList(
        oldListId: Int,
        newListId: Int,
        ids: MutableCollection<Int>,
    ): Boolean {
        return mediaListRepository.moveMediaToList(oldListId, newListId, ids)
    }

    suspend fun addMediumToList(listId: Int, ids: MutableCollection<Int>): Boolean {
        return MediaListRepository.getInstance(getApplication()).addMediumToList(listId, ids)
    }
}